editor/utils/heartbeat.js

```diff
@@ -0,0 +1,173 @@
+import store from '../store';
+import {
+	getCurrentPost,
+	getEditedPostTitle,
+	getEditedPostExcerpt,
+	getEditedPostContent,
+} from '../store/selectors';
+import {
+	toggleAutosave,
+	removeNotice,
+	toggleNetworkIsConnected,
+	showDisconnectionNotice,
+} from '../store/actions';
+
+import { compact } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { doAction } from '@wordpress/hooks';
+
+const DISCONNECTION_NOTICE_ID = 'DISCONNECTION_NOTICE_ID';
+
+export function setupHeartbeat() {
+	const $document = jQuery( document );
+	/**
+	 * Configure heartbeat to refresh the wp-api nonce, keeping the editor authorization intact.
+	 *
+	 * @todo update the _wpnonce used by autosaves.
+	 */
+	$document.on( 'heartbeat-tick', ( event, response ) => {
+		if ( response[ 'rest-nonce' ] ) {
+			window.wpApiSettings.nonce = response[ 'rest-nonce' ];
+		}
+	} );
+
+	/**
+	 * Concatenate the title, content and excerpt.
+	 *
+	 * This is used to track changes when auto-saving.
+	 *
+	 * @since 1.9.0
+	 *
+	 * @param {Object} state The current state.
+	 *
+	 * @returns {string} A concatenated string with title, content and excerpt.
+	 */
+	const getCompareString = function( state ) {
+		return compact( [
+			getEditedPostTitle( state ),
+			getEditedPostContent( state ),
+			getEditedPostExcerpt( state ),
+		] ).join( '::' );
```

Comment:
Can't we use a flag like `isAutoSaveDirty` or something instead of this. It seems a bit fragile as it doesn't account for all the potential changes. I think we can use the `withChangeDetection` Helper to add a `isAutoSaveDirty`. We'd have to update the `withChangeDetection` helper to make the "isDirty" flag name customizable) cc @aduth

Reply:
that would make sense if possible. Originally I was tracking the autosaves in Redux, however since its currently only used here in heartbeat, I wasn't sure we needed to do that.

Reply:
@youknowriad I worked on adding this, storing the autosave when heartbeat fires and using withChangeDetection and isDirty to determine if its changed. let me know if this looks better